django/db/migrations/state.py

```diff
@@ -188,6 +193,89 @@ def _reload(self, related_models):
         # Render all models
         self.apps.render_multiple(states_to_be_rendered)
 
+    def resolve_fields_and_relations(self):
+        for model_state in self.models.values():
+            for field_name, field in model_state.fields:
+                field.name = field_name
+                field.model = "{}.{}".format(model_state.app_label, model_state.name)
+                if field.remote_field and field.remote_field.model:
+                    self.fill_remote_field_name(field)
+
+        self.relations = defaultdict(lambda: defaultdict(list))
+        concretes, proxies = self._get_concrete_models_mapping_and_proxy_models()
+
+        real_apps = set(self.real_apps)
+        for model_key in concretes:
+            model_state = self.models[model_key]
+            for field_name, field in model_state.fields:
+                remote_field = field.remote_field
+                if remote_field:
+                    remote_model_key = resolve_model_key(*model_key, remote_field.model)
+                    if remote_model_key[0] not in real_apps and remote_model_key in concretes:
+                        remote_model_key = concretes[remote_model_key]
+                    self.relations[remote_model_key][model_key].append((field_name, field))
+
+                    through = getattr(remote_field, 'through', None)
+                    if through and not model_state.options.get('auto_created'):
+                        through_model_key = resolve_model_key(*model_key, through)
+                        if through_model_key[0] not in real_apps:
+                            through_model_key = concretes[through_model_key]
+                        self.relations[through_model_key][model_key].append((field_name, field))
+        for model_key in proxies:
+            self.relations[model_key] = self.relations[concretes[model_key]]
+
+    def fill_remote_field_name(self, field):
+        if self.is_swappable_model(field.remote_field.model):
+            field.remote_field.field_name = "id"
+        else:
+            try:
+                model_state = self.get_concrete_model_state(field.remote_field.model)
+                base = model_state.bases[0]
+                if not isinstance(base, str) and base.__name__ == "Model":
+                    field.remote_field.field_name = model_state.get_primary_key_field_name()
+                    model_state.all_fields.add(field.remote_field)
+                else:
+                    base_model_state = self.get_concrete_model_state(base)
+                    field.remote_field.field_name = '%s_ptr' % base_model_state.name_lower
+                    base_model_state.all_fields.add(field.remote_field)
+            except KeyError:
+                pass  # not migrated
```

Comment:
I'd be great to have bit more clarity about which failure case we're handling here with `KeyError`. This seems like a complex routine wrapped in a catch all.

Reply:
The 3 tests that fail because we are not able to resolve the concrete model are:
* `test_first_dependency`
* `test_last_dependency`
* `test_swappable_changed`

Basically because the FK fields point a model that does not exist in the state `"migrations.UnmigratedModel"` (`https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L352`) or the model is inhering from an unknown model in the state `AbstractBaseUse` (https://github.com/django/django/blob/master/tests/migrations/test_autodetector.py#L441-L444)

These can probably be handled more elegantly